apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/dynamic/ConfigurationDiscoveryService.java

```diff
@@ -0,0 +1,241 @@
+                        ));
+                    } else {
+                        // Don't need to notify, stay in the same config value.
+                    }
+                }
+            } else {
+                LOGGER.warn("Config {} from OAP, doesn't match any watcher, ignore.", propertyKey);
+            }
+        });
+        this.uuid = responseUuid;
+
+        LOGGER.trace("Current configurations after the sync." + LINE_SEPARATOR + register.toString());
+    }
+
+    /**
+     * get agent dynamic config through gRPC.
+     */
+    private void getAgentDynamicConfig() {
+        LOGGER.debug("ConfigurationDiscoveryService running, status:{}.", status);
+
+        if (GRPCChannelStatus.CONNECTED.equals(status)) {
+            try {
+                ConfigurationSyncRequest.Builder builder = ConfigurationSyncRequest.newBuilder();
+                builder.setService(Config.Agent.SERVICE_NAME);
+                if (null != uuid) {
+                    builder.setUuid(uuid);
+                }
+
+                if (configurationDiscoveryServiceBlockingStub != null) {
+                    final Commands commands = configurationDiscoveryServiceBlockingStub.withDeadlineAfter(
+                        GRPC_UPSTREAM_TIMEOUT, TimeUnit.SECONDS
+                    ).fetchConfigurations(builder.build());
+                    ServiceManager.INSTANCE.findService(CommandService.class).receiveCommand(commands);
+                }
+            } catch (Throwable t) {
+                LOGGER.error(t, "ConfigurationDiscoveryService execute fail.");
+                ServiceManager.INSTANCE.findService(GRPCChannelManager.class).reportError(t);
+            }
+        }
+    }
+
+    /**
+     * Local dynamic configuration center.
+     */
+    public class Register {
+        private Map<String, WatcherHolder> register = new HashMap<>();
+
+        private boolean containsKey(String key) {
+            return register.containsKey(key);
+        }
+
+        private void put(String key, WatcherHolder holder) {
+            register.put(key, holder);
+        }
+
+        public WatcherHolder get(String name) {
+            return register.get(name);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder registerTableDescription = new StringBuilder();
+            registerTableDescription.append("Following dynamic config property are available.").append(LINE_SEPARATOR);
+            registerTableDescription.append("---------------------------------------------").append(LINE_SEPARATOR);
+            register.forEach((key, holder) -> {
+                AgentConfigChangeWatcher watcher = holder.getWatcher();
+                registerTableDescription.append("key:")
+                                        .append(key)
+                                        .append("    value(current):")
+                                        .append(watcher.value())
+                                        .append(LINE_SEPARATOR);
```

Comment:
We don't need to format the logs, no one is watching the logs for details. There are many tool to scratch, format, and visualize the logs. It isn't our reposonsibility.

Reply:
done.